### PR TITLE
Telcodocs-1010: Create Pre-flight and Pre-build content for KMMO

### PR DESCRIPTION
**Pull request by StephenJamesSmith**
Create Pre-flight and Pre-build content for KMMO

OCP version: openshift-4.12.z

Version: [KMMO 1.0](https://issues.redhat.com/issues/?jql=project+%3D+TELCODOCS+AND+fixVersion+%3D+%22KMMO+1.0%22)

Issue: https://issues.redhat.com/browse/TELCODOCS-1010

Link to docs preview: https://56634--docspreview.netlify.app/openshift-enterprise/latest/updating/kmm-preflight-validation.html

SME review: [@quba](https://redhat-internal.slack.com/team/U02GW51L0GZ)
T&PS: @slovern 
Dev @yevgeny-shnaidman
QE review: [@Costi](https://redhat-internal.slack.com/team/ULP4MT9GB)

### CODE REVIEW

File: modules/kmm-example-cr.adoc [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+//
+// * updating/kmm-preflight-validation.adoc
+
+:_content-type: CONCEPT
+[id="kmm-example-cr_{context}"]
+= Example CR
+
+This section shows an example of the `PreflightValidationOCP` resource in the YAML format.
+
+The example verifies all the currently present modules against the upcoming kernel version including in the
```
Comment (cdvultur) [inline modules/kmm-example-cr.adoc:11]: kernel version **included** in the OCP

**Review by cdvultur — COMMENTED** [review]
Overall this looks good. 

I would just add a note on the fact that the build section of the module should handle properly the kernel version. 

@yevgeny-shnaidman How do you think we should specify this in the docs?

Good way : `RUN depmod -b /opt ${KERNEL_VERSION}`
Bad way: `RUN depmod -b /opt`

**Comment by ggordaniRed** [comment]
looking good to me

**Comment by StephenJamesSmith** [comment]
[@Costi](https://redhat-internal.slack.com/team/ULP4MT9GB) Are you ready to review?

File: modules/kmm-example-cr.adoc (hunk 1 above)
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:11]: changed.

**Comment by StephenJamesSmith** [comment]
@yevgeny-shnaidman [@Costi](https://redhat-internal.slack.com/team/ULP4MT9GB) Please review and provide any comments. If the content is satisfactory, please add /LGTM in a comment.

**Comment by StephenJamesSmith** [comment]
@quba @yevgeny-shnaidman @Costi All: Please review and provide your comments, if you have not already done so. If the content is satisfactory, please add the comment /LGTM. I can't send this PR to OCP Peer Review without your signoffs. Thanks!

**Comment by StephenJamesSmith** [comment]
@yevgeny-shnaidman - Please /lgtm

File: modules/kmm-build-validation-stage.adoc [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+:_content-type: CONCEPT
+[id="kmm-build-validation-stage_{context}"]
+= Build validation stage
+
+Build validation is executed only when image validation has failed and there is a `build` section in the `Module` that is relevant for the upgraded kernel. Build validation attempts to run the build job and validate that it finishes successfully.
+
+[NOTE]
+====
+You need to specify the kernel version when running `depmod`, as shown here:
```
Comment (adellape) [inline modules/kmm-build-validation-stage.adoc:13]: s/need to/must/

File: modules/kmm-build-validation-stage.adoc [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+[source,terminal]
+----
+$ RUN depmod -b /opt ${KERNEL_VERSION}.
```
Comment (adellape) [inline modules/kmm-build-validation-stage.adoc:16]: Just checking whether this `.` on the end is supposed to be there.

File: modules/kmm-build-validation-stage.adoc [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+----
+====
+
+If the `PushBuiltImage` flag is defined in the `PreflightValidationOCP` CR, it will also try to push the resulting image
+into its repo. The resulting image name is taken from the definition of the `containerImage` field of the `Module` CR.
```
Comment (adellape) [inline modules/kmm-build-validation-stage.adoc:21]: s/repo/repository/

Comment (adellape) [inline modules/kmm-build-validation-stage.adoc:21]: For first appearance: "custom resource (CR)"

File: modules/kmm-build-validation-stage.adoc [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+[NOTE]
+====
+If the `sign` section is defined for the upgraded kernel, then the resulting image will not be `containerImage`
```
Comment (adellape) [inline modules/kmm-build-validation-stage.adoc:25]: "the `containerImage`"

File: modules/kmm-example-cr.adoc [hunk 6]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+//
+// * updating/kmm-preflight-validation.adoc
+
+:_content-type: CONCEPT
+[id="kmm-example-cr_{context}"]
+= Example CR
```
Comment (adellape) [inline modules/kmm-example-cr.adoc:7]: Suggest a more specific title, e.g. "Example PreflightValidationOCP resource"

File: modules/kmm-sign-validation-stage.adoc [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+[id="kmm-sign-validation-stage_{context}"]
+= Sign validation stage
+
+Sign validation is executed only when image validation has failed, there is a `sign` section in the `Module` that is relevant for the upgrade kernel, and build validation finished successfully in the event there was a `build` section in the `Module` relevant for the upgraded kernel.
+Sign validation will try to run the sign job and validate that it finishes successfully.
+
+If the `PushBuiltImage` flag is defined in the `PreflightValidationOCP` CR, sign validation will also try to
+push the resulting image to its registry.
+
+The resulting image is always the image defined in the `ContainerImage` field of the `Module`. The input image is either the output of the Build stage, or an image defined in the `UnsignedImage` field.
```
Comment (adellape) [inline modules/kmm-sign-validation-stage.adoc:15]: Should it be lowercase "c" `containerImage` like the others?

File: modules/kmm-validation-kickoff.adoc [hunk 8]
```diff
@@ -0,0 +1,30 @@
+// Module included in the following assemblies:
+//
+// * updating/kmm-preflight-validation.adoc
+
+:_content-type: CONCEPT
+[id="kmm-validation-kickoff_{context}"]
+= Validation kickoff
+
+Preflight validation is triggered by creating a `PreflightValidationOCP` resource in the cluster. This Spec contains two
```
Comment (adellape) [inline modules/kmm-validation-kickoff.adoc:9]: s/Spec/spec/

File: modules/kmm-validation-kickoff.adoc [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+	// +optional
+	PushBuiltImage bool `json:"pushBuiltImage"` <2>
+}
+----
+
+<1> `ReleaseImage` - Mandatory field that provides the name of the release image for the
+{product-title} version the cluster is upgraded to.
+
+<2> `PushBuiltImage` - If `true`, then the images created during the Build and Sign validation
+will be pushed to their repositories (`false` by default).
```
Comment (adellape) [inline modules/kmm-validation-kickoff.adoc:30]: s/will be/are/

File: modules/kmm-validation-lifecycle.adoc [hunk 10]
```diff
@@ -0,0 +1,11 @@
+// Module included in the following assemblies:
+//
+// * updating/kmm-preflight-validation.adoc
+
+:_content-type: CONCEPT
+[id="kmm-validation-lifecycle_{context}"]
+= Validation lifecycle
+
+Preflight validation attempts to validate every module loaded in the cluster. Preflight will stop running validation on a `Module`, once the validation is successful.
```
Comment (adellape) [inline modules/kmm-validation-lifecycle.adoc:9]: No comma and s/once/after/:

"on a `Module` resource after"

And either "`Module` resource" (or whatever is accurate) if using the marked-up uppercase, or just "module" for a generic reference.

File: updating/kmm-preflight-validation.adoc [hunk 11]
```diff
@@ -0,0 +1,22 @@
+:_content-type: ASSEMBLY
+[id="kmm-preflight-validation"]
+= Preflight validation for Kernel Module Management (KMM) Modules
+include::_attributes/common-attributes.adoc[]
+:context: kmm-preflight-validation
+
+toc::[]
+
+Before performing an upgrade on the cluster with applied KMM modules, the administrator needs to verify
```
Comment (adellape) [inline updating/kmm-preflight-validation.adoc:9]: s/needs to/must/

File: updating/kmm-preflight-validation.adoc [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+that installed kernel modules (via KMM) are able to be installed on the nodes after the cluster upgrade and
```
Comment (adellape) [inline updating/kmm-preflight-validation.adoc:10]: See IBM SG for word usage on "via" and avoiding using parentheses in general text.

**Comment by adellape** [comment]
@StephenJamesSmith Also, which OCP version branches should this PR apply to?

File: modules/kmm-build-validation-stage.adoc (hunk 3 above)
Comment (StephenJamesSmith) [reply modules/kmm-build-validation-stage.adoc:16]: It's not. I'll remove it. Thx.

File: modules/kmm-build-validation-stage.adoc (hunk 4 above)
Comment (StephenJamesSmith) [reply modules/kmm-build-validation-stage.adoc:21]: changed.

File: modules/kmm-build-validation-stage.adoc [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+field of the `Module` CR, but a temporary image name, since the resulting image should be the product of Sign flow.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-build-validation-stage.adoc:26]: changed.

File: modules/kmm-example-cr.adoc (hunk 6 above)
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:7]: changed.

File: modules/kmm-example-cr.adoc [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+This section shows an example of the `PreflightValidationOCP` resource in the YAML format.
+
+The example verifies all the currently present modules against the upcoming kernel version included in the
+OCP release `4.11.18`, which the release image `quay.io/openshift-release-dev/ocp-release@sha256:22e149142517dfccb47be828f012659b1ccf71d26620e6f62468c264a7ce7863`
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-example-cr.adoc:12]: changed both.

File: modules/kmm-image-validation-stage.adoc [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+Image validation is always the first stage of the preflight validation to be executed.
+If image validation is successful, no other validations are run on that specific module.
+
+Image validation consists of two stages:
+
+. Image existence and accessibility. The code tries to access the image defined for the upgraded kernel
+in the module and get its manifests.
+
+. Verify the presence of the kernel module defined in the `Module` in the correct path for future
+`modprobe` execution. The correct path is `<DirName>/lib/modules/<UpgradedKernel>/`.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-image-validation-stage.adoc:18]: changed

File: modules/kmm-image-validation-stage.adoc [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+If this validation is successful, it probably means that the kernel module was compiled with the correct linux headers.
```
[the remark this answers is not in the document]
Comment (StephenJamesSmith) [reply modules/kmm-image-validation-stage.adoc:20]: changed

File: modules/kmm-sign-validation-stage.adoc (hunk 7 above)
Comment (StephenJamesSmith) [reply modules/kmm-sign-validation-stage.adoc:15]: yes. changed.

File: modules/kmm-validation-kickoff.adoc (hunk 8 above)
Comment (StephenJamesSmith) [reply modules/kmm-validation-kickoff.adoc:9]: changed

File: updating/kmm-preflight-validation.adoc (hunk 12 above)
Comment (StephenJamesSmith) [reply updating/kmm-preflight-validation.adoc:10]: Changed to "kernel modules installed using KMM are..."

**Pull request merged**